src/Workspaces/Core/Portable/Workspace/Solution/MetadataOnlyImage.cs

```diff
@@ -97,18 +102,15 @@ public MetadataReference CreateReference(ImmutableArray<string> aliases, bool em
                 // give them pointer to the native memory. also we need to handle lifetime ourselves.
                 var metadata = AssemblyMetadata.Create(ModuleMetadata.CreateFromImage(supportNativeMemory.GetPointer(), (int)stream.Length));
 
-                var referenceWithNativeMemory = metadata.GetReference(
+                // Tie lifetime of stream to metadata we created. It is important to tie this to the Metadata and not the
+                // metadata reference, as PE symbols hold onto just the Metadata.
+                s_lifetime.Add(metadata, supportNativeMemory);
```

Comment:
Indeed, we created a new object a few lines above, so there can't be a race here. I'm switching it to an Add to save the cost of the closure.